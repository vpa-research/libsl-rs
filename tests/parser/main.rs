use std::fs;

use insta::{assert_ron_snapshot, Settings};
use libsl::{LibSl, WithLibSl};
use pathdiff::diff_paths;

#[test]
fn test_snapshots() {
    let mut settings = Settings::clone_current();
    settings.set_omit_expression(true);
    let _guard = settings.bind_to_scope();

    insta::glob!("test-data/**/*.libsl", |path| {
        let contents = fs::read_to_string(path)
            .map_err(|e| format!("could not read `{}` to string: {e}", path.display()))
            .unwrap();

        let mut settings = Settings::clone_current();
        settings.set_description(format!("// Input file:\n{contents}"));
        let _guard = settings.bind_to_scope();

        let mut libsl = LibSl::new();
        let ast = libsl.parse_file(
            diff_paths(path, env!("CARGO_MANIFEST_DIR")).unwrap().display().to_string(),
            &contents
        );

        assert_ron_snapshot!(match &ast {
            Ok(file) => Ok(file.with_libsl(&libsl)),
            Err(e) => Err(e.to_string()),
        });
    });
}
