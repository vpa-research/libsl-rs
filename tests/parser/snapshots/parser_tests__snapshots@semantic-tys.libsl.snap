---
source: tests/parser/main.rs
description: "// Input file:\nlibsl \"1.0.0\";\n\nlibrary semantic_tys;\n\n// empty.\ntypes {}\n\ntypes {\n    Simple(RealTy);\n\n    Enumerated(RealTy) {\n        foo: 42;\n        bar: 24;\n    }\n}\n"
input_file: tests/parser/test-data/semantic-tys.libsl
---
Ok(File(
  loc: Span(
    start: 0,
    len: 154,
    file: "tests/parser/test-data/semantic-tys.libsl",
    line: Some(1),
    col: None,
  ),
  header: Some(Header(
    loc: Span(
      start: 0,
      len: 36,
      file: "tests/parser/test-data/semantic-tys.libsl",
      line: Some(1),
      col: None,
    ),
    libsl_version: "1.0.0",
    library_name: "semantic_tys",
    version: None,
    language: None,
    url: None,
  )),
  decls: [
    Decl(
      loc: Span(
        start: 71,
        len: 14,
        file: "tests/parser/test-data/semantic-tys.libsl",
        line: Some(9),
        col: Some(4),
      ),
      kind: SemanticTy(DeclSemanticTy(
        annotations: [],
        ty_name: QualifiedTyName(
          ty_name: FullName(
            components: [
              Name(
                loc: Span(
                  start: 71,
                  len: 5,
                  file: "tests/parser/test-data/semantic-tys.libsl",
                  line: Some(9),
                  col: Some(4),
                ),
                name: "Simple",
              ),
            ],
          ),
          generics: [],
        ),
        real_ty: TyExpr(
          loc: Span(
            start: 78,
            len: 5,
            file: "tests/parser/test-data/semantic-tys.libsl",
            line: Some(9),
            col: Some(11),
          ),
          kind: Name(TyExprName(
            ty_name: FullName(
              components: [
                Name(
                  loc: Span(
                    start: 78,
                    len: 5,
                    file: "tests/parser/test-data/semantic-tys.libsl",
                    line: Some(9),
                    col: Some(11),
                  ),
                  name: "RealTy",
                ),
              ],
            ),
            generics: [],
          )),
        ),
        kind: Simple,
      )),
    ),
    Decl(
      loc: Span(
        start: 92,
        len: 59,
        file: "tests/parser/test-data/semantic-tys.libsl",
        line: Some(11),
        col: Some(4),
      ),
      kind: SemanticTy(DeclSemanticTy(
        annotations: [],
        ty_name: QualifiedTyName(
          ty_name: FullName(
            components: [
              Name(
                loc: Span(
                  start: 92,
                  len: 9,
                  file: "tests/parser/test-data/semantic-tys.libsl",
                  line: Some(11),
                  col: Some(4),
                ),
                name: "Enumerated",
              ),
            ],
          ),
          generics: [],
        ),
        real_ty: TyExpr(
          loc: Span(
            start: 103,
            len: 5,
            file: "tests/parser/test-data/semantic-tys.libsl",
            line: Some(11),
            col: Some(15),
          ),
          kind: Name(TyExprName(
            ty_name: FullName(
              components: [
                Name(
                  loc: Span(
                    start: 103,
                    len: 5,
                    file: "tests/parser/test-data/semantic-tys.libsl",
                    line: Some(11),
                    col: Some(15),
                  ),
                  name: "RealTy",
                ),
              ],
            ),
            generics: [],
          )),
        ),
        kind: Enumerated([
          SemanticTyEnumValue(
            name: Name(
              loc: Span(
                start: 121,
                len: 2,
                file: "tests/parser/test-data/semantic-tys.libsl",
                line: Some(12),
                col: Some(8),
              ),
              name: "foo",
            ),
            expr: Expr(
              loc: Span(
                start: 126,
                len: 1,
                file: "tests/parser/test-data/semantic-tys.libsl",
                line: Some(12),
                col: Some(13),
              ),
              kind: PrimitiveLit(ExprPrimitiveLit(
                lit: Int(I32(42)),
              )),
            ),
          ),
          SemanticTyEnumValue(
            name: Name(
              loc: Span(
                start: 138,
                len: 2,
                file: "tests/parser/test-data/semantic-tys.libsl",
                line: Some(13),
                col: Some(8),
              ),
              name: "bar",
            ),
            expr: Expr(
              loc: Span(
                start: 143,
                len: 1,
                file: "tests/parser/test-data/semantic-tys.libsl",
                line: Some(13),
                col: Some(13),
              ),
              kind: PrimitiveLit(ExprPrimitiveLit(
                lit: Int(I32(24)),
              )),
            ),
          ),
        ]),
      )),
    ),
  ],
))
